src/storage/src/storage/checksum.rs

```diff
@@ -0,0 +1,539 @@
+        assert!(matches!(&err, &ChecksumMismatch::Both { .. }), "{err:?}");
+    }
+
+    fn empty() -> bytes::Bytes {
+        bytes::Bytes::new()
+    }
+
+    fn data() -> bytes::Bytes {
+        bytes::Bytes::from_static(b"the quick brown fox jumps over the lazy dog")
+    }
+
+    #[test]
+    fn null() {
+        let mut engine = Null;
+        engine.update(0, &data());
+        assert_eq!(engine.finalize(), ObjectChecksums::new());
+    }
+
+    #[test]
+    fn precomputed() {
+        let mut engine = Precomputed;
+        engine.update(0, &data());
+        assert_eq!(engine.finalize(), ObjectChecksums::new());
+    }
+
+    #[test_case(empty())]
+    #[test_case(data())]
+    fn crc32c_basic(input: bytes::Bytes) {
+        let mut engine = Crc32c::new();
+        engine.update(0, &input);
+        let want = crc32c::crc32c(&input);
+        assert_eq!(engine.finalize(), ObjectChecksums::new().set_crc32c(want));
+    }
+
+    #[test]
+    fn crc32c_in_parts() {
+        let input = data();
+
+        let mut engine = Crc32c::new();
+        engine.update(0, &input.slice(0..4));
+        engine.update(0, &input.slice(0..4));
+        engine.update(4, &input.slice(4..8));
+        engine.update(6, &input.slice(6..12));
+        engine.update(8, &input.slice(8..));
+        let want = crc32c::crc32c(&data());
+        assert_eq!(engine.finalize(), ObjectChecksums::new().set_crc32c(want));
+    }
+
+    #[test_case(empty())]
+    #[test_case(data())]
+    fn md5_basic(input: bytes::Bytes) {
+        let mut engine = Md5::new();
+        engine.update(0, &input);
+        let digest = md5::compute(&input);
+        let want = bytes::Bytes::from_owner(Vec::from_iter(digest.0));
+        assert_eq!(engine.finalize(), ObjectChecksums::new().set_md5_hash(want));
+    }
+
+    #[test]
+    fn md5_in_parts() {
+        let input = data();
+        let mut engine = Md5::new();
+        let digest = md5::compute(&input);
+        let want = bytes::Bytes::from_owner(Vec::from_iter(digest.0));
+
+        engine.update(0, &input.slice(0..4));
+        engine.update(0, &input.slice(0..4));
+        engine.update(4, &input.slice(4..8));
+        engine.update(6, &input.slice(6..12));
+        engine.update(8, &input.slice(8..));
+        assert_eq!(engine.finalize(), ObjectChecksums::new().set_md5_hash(want));
+    }
+
+    #[test]
+    fn md5_and_crc32_in_parts() {
+        let input = data();
+        let mut engine = Md5::from_inner(Crc32c::new());
+        let digest = md5::compute(&input);
+        let md5_want = bytes::Bytes::from_owner(Vec::from_iter(digest.0));
+        let crc32_want = crc32c::crc32c(&input);
+
+        engine.update(0, &input.slice(0..4));
+        engine.update(0, &input.slice(0..4));
+        engine.update(4, &input.slice(4..8));
+        engine.update(6, &input.slice(6..12));
+        engine.update(0, &input.slice(0..4));
+        engine.update(8, &input.slice(8..));
+        engine.update(100, &input.slice(0..));
```

Comment:
comment: this is kinda wacky. I think you are testing an update past the size of the data. Maybe remove or explain what this is for?